src/mlir/lowering.md

```diff
@@ -0,0 +1,517 @@
+storage until we can lower them into LLVM. LLVM IR supports string literals and
+we can call string maniupulation library functions to operate on them. The MLIR
+Toy example has a mock implementation of that process.
+
+Example:
+```
+  // Simple string
+  %simple_string = verona.constant "This is a string" : !verona.string
+
+  // Interpolated string ("My name is $1 and I'm $2 years old", name, age)
+  %0 = verona.constant "My name is " : !verona.string
+  %1 = verona.constant " and I'm " : !verona.string
+  %2 = verona.constant " years old" : !verona.string
+
+  // This is still not decided, but just for illustration purposes
+  // We could also pass this to the representation, so it can do its own thing
+  %name_str = verona.call "toString" [ %name ] : !verona.string
+  %age_str = verona.call "toString" [ %age ] : !verona.string
+
+  // This is whatever runtime string representation we choose
+  // assuming we have a constructor with a list of literal strings
+  %utf8 = verona.new_object @UTF8String [%0, %name_str, %1, %age_str, %2]
+```
+
+In the case above, if all variables are already compile-time constant strings
+at the time of lowering, the compiler has the ability to fuse the strings,
+creating a simple string (or at least less splits).
+
+### UnitStmt
+
+No idea.
+
+### MatchBindStmt
+
+An unchecked cast for runtime `match` statements. If the type of the runtime
+object isn't compatible with the cast, an invalid object returns?
+
+A simple way to lower `match` statements is to chain basic blocks with
+conditional branches.
+
+```
+  %val = the runtime value passed to the `match` statement
+  %0, %val0 = verona.match @Type1 [ %val ]
+  cond_br %0, ^bb1(%val0), ^bb2
+^bb1(%local_val : Type1):
+  // Implement `case Type1`
+  br ^lastbb
+
+^bb2:
+  %1, %val1 = verona.match @Type2 [ %val ]
+  cond_br %1, ^bb3(%val1), ^bb4
+^bb3(%local_val : Type2):
+  // Implement `case Type2`
+  br ^lastbb
+
+  ...
+
+^bbX:
+  %y, %valy = verona.match @TypeY [ %val ]
+  cond_br %y, ^bbY(%valy), ^lastbb
+^bbY(%local_val : TypeY):
+  // Implement `case TypeY`
+  br ^lastbb
+
+^lastbb: // no args
+
+```
+
+LLVM IR has support for `switch` statements, but they require the
+type of the object to be an integer. We could create a dispatch table and use
+offset in that table as the switch's value. But this would be an optimisation
+that has no clear benefit for now.
+
+### ViewStmt
+
+Created a mutable reference to an object. The validity of the type system has
+already been checked by the compiler at this point, so all reads and writes to
+those values are valid.
+
+At the 2IR level, there is no need to represent this statement as an operation
+and the only implementation detail is to add the previous value to the new
+symbol.
+
+### EndScopeStmt / OverwriteStmt
+
+When variables go out of scope, either lexical or type related (ex. move
+semantics). We need to call `release` on owned region types (region, iso, imm)
+but not on unowned types (mut).
```

Comment:
It's more complicated than types. It's about *values*. If you have a variable of type `iso | mut | imm`, there are 4 possible ways to deal with it (`{Cown,Immutable,Region}::release`, or nothing), and you cannot pick the correct one until runtime.

Also "owned region types (region, iso, imm)" is confusing. `imm` is not a region.

Reply:
Isn't this just a runtime `match`? I'm expecting that all runtime unions will need runtime `match` statements. It should be fine for either 1IR or 2IR to add the calls depending on which level we're lowering.

Fixed the "owned types" confusion.